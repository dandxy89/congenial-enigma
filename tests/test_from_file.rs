use std::path::PathBuf;

use lp_parser_rs::{
    model::{lp_error::LPParserError, lp_problem::LPProblem},
    parse::{parse_file, parse_lp_file},
};

#[macro_export]
macro_rules! generate_test {
    ($test_name:ident, $file:expr, $o_sum:expr, $c_sum:expr) => {
        #[test]
        fn $test_name() {
            let result = read_file_from_resources($file).unwrap();

            #[cfg(feature = "serde")]
            insta::assert_yaml_snapshot!(result, {
                ".variables" => insta::sorted_redaction(),
                ".objectives" => insta::sorted_redaction(),
                ".constraints" => insta::sorted_redaction(),
            });

            let summation = &result.objectives.iter().map(|o| o.coefficients.iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($o_sum, *summation, abs <= 1e-2);

            let summation = &result.constraints.iter().map(|(_, c)| c.coefficients().iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($c_sum, *summation, abs <= 1e-2);
        }
    };
}

generate_test!(afiro, "afiro.lp", -4.100, 25.369);
generate_test!(afiro_ext, "afiro_ext.lp", -5.239, 25.369);
generate_test!(boeing1, "boeing1.lp", 1187.985, 194612.346);
generate_test!(boeing2, "boeing2.lp", 78.488, 20863.836);
generate_test!(fit1d, "fit1d.lp", 82457.0, -146871.180);
generate_test!(kb2, "kb2.lp", 11.675, 10143.724);
generate_test!(pulp, "pulp.lp", 1., 73.44);
generate_test!(pulp2, "pulp2.lp", -6.0, 147.);
generate_test!(sc50a, "sc50a.lp", -1., 30.3);
generate_test!(model2, "model2.lp", 0., 6.);
generate_test!(limbo, "limbo.lp", 2., 0.);
generate_test!(obj3_2cons, "3obj_2cons.lp", 1., -186.);
generate_test!(obj_2cons_only_binary_vars, "2obj_2cons_only_binary_vars.lp", -7.5, -186.);
generate_test!(obj_2cons_all_variable_types, "2obj_2cons_all_variable_types.lp", -7.5, -186.);
generate_test!(obj_1cons_all_variables_with_bounds, "1obj_1cons_all_variables_with_bounds.lp", -1., 16.5);
generate_test!(semi_continuous, "semi_continuous.lp", 2., 0.);
generate_test!(sos, "sos.lp", 0., 17.5);
generate_test!(test, "test.lp", 2., 2.9899);
generate_test!(test2, "test2.lp", -6., 147.);
generate_test!(empty_bounds, "empty_bounds.lp", 11., 2.);
generate_test!(blank_lines, "blank_lines.lp", 11., 2.);
generate_test!(optional_labels, "optional_labels.lp", 11., 1.);
generate_test!(infile_comments, "infile_comments.lp", 43., 7.);
generate_test!(infile_comments2, "infile_comments2.lp", 43., 0.);
generate_test!(missing_signs, "missing_signs.lp", 43., -3.);
generate_test!(fit2d, "fit2d.lp", 349048.9, -296677.389);
generate_test!(scientific_notation, "scientific_notation.lp", 238.0324, -43.87680);

#[test]
fn invalid() {
    let result = read_file_from_resources("invalid.lp");
    assert!(result.is_err());
}

fn read_file_from_resources(file_name: &str) -> Result<LPProblem, LPParserError> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;
    parse_lp_file(&contents)
}
