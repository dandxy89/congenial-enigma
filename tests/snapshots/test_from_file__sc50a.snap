---
source: tests/test_from_file.rs
expression: result
---
problem_name: sc50a.lp
problem_sense: Minimize
variables:
  COL00001: General
  COL00002: General
  COL00003: General
  COL00004: General
  COL00005: General
  COL00006: General
  COL00007: General
  COL00008: General
  COL00009: General
  COL00010: General
  COL00011: General
  COL00012: General
  COL00013: General
  COL00014: General
  COL00015: General
  COL00016: General
  COL00017: General
  COL00018: General
  COL00019: General
  COL00020: General
  COL00021: General
  COL00022: General
  COL00023: General
  COL00024: General
  COL00025: General
  COL00026: General
  COL00027: General
  COL00028: General
  COL00029: General
  COL00030: General
  COL00031: General
  COL00032: General
  COL00033: General
  COL00034: General
  COL00035: General
  COL00036: General
  COL00037: General
  COL00038: General
  COL00039: General
  COL00040: General
  COL00041: General
  COL00042: General
  COL00043: General
  COL00044: General
  COL00045: General
  COL00046: General
  COL00047: General
  COL00048: General
objectives:
  - name: MAXIM
    coefficients:
      - var_name: COL00004
        coefficient: -1
constraints:
  "ROW00001:":
    Standard:
      name: "ROW00001:"
      coefficients:
        - var_name: COL00001
          coefficient: 2
        - var_name: COL00002
          coefficient: 1
        - var_name: COL00003
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 170
  "ROW00002:":
    Standard:
      name: "ROW00002:"
      coefficients:
        - var_name: COL00001
          coefficient: 1
        - var_name: COL00002
          coefficient: 2
        - var_name: COL00003
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 130
  "ROW00004:":
    Standard:
      name: "ROW00004:"
      coefficients:
        - var_name: COL00004
          coefficient: 1
        - var_name: COL00005
          coefficient: -1
      sense: Equal
      rhs: 0
  "ROW00005:":
    Standard:
      name: "ROW00005:"
      coefficients:
        - var_name: COL00001
          coefficient: -1
        - var_name: COL00006
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00006:":
    Standard:
      name: "ROW00006:"
      coefficients:
        - var_name: COL00002
          coefficient: -1
        - var_name: COL00007
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00007:":
    Standard:
      name: "ROW00007:"
      coefficients:
        - var_name: COL00003
          coefficient: -1
        - var_name: COL00008
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00008:":
    Standard:
      name: "ROW00008:"
      coefficients:
        - var_name: COL00006
          coefficient: -1
        - var_name: COL00009
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00009:":
    Standard:
      name: "ROW00009:"
      coefficients:
        - var_name: COL00007
          coefficient: -1
        - var_name: COL00010
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00010:":
    Standard:
      name: "ROW00010:"
      coefficients:
        - var_name: COL00008
          coefficient: -1
        - var_name: COL00011
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00011:":
    Standard:
      name: "ROW00011:"
      coefficients:
        - var_name: COL00009
          coefficient: -1
        - var_name: COL00012
          coefficient: 2
        - var_name: COL00013
          coefficient: 1
        - var_name: COL00014
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 170
  "ROW00012:":
    Standard:
      name: "ROW00012:"
      coefficients:
        - var_name: COL00010
          coefficient: -1
        - var_name: COL00012
          coefficient: 1
        - var_name: COL00013
          coefficient: 2
        - var_name: COL00014
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 130
  "ROW00013:":
    Standard:
      name: "ROW00013:"
      coefficients:
        - var_name: COL00005
          coefficient: 1
        - var_name: COL00011
          coefficient: -1
      sense: LessOrEqual
      rhs: 0
  "ROW00014:":
    Standard:
      name: "ROW00014:"
      coefficients:
        - var_name: COL00004
          coefficient: 1.1
        - var_name: COL00015
          coefficient: -1
      sense: Equal
      rhs: 0
  "ROW00015:":
    Standard:
      name: "ROW00015:"
      coefficients:
        - var_name: COL00005
          coefficient: 1
        - var_name: COL00015
          coefficient: 1
        - var_name: COL00016
          coefficient: -1
      sense: Equal
      rhs: 0
  "ROW00016:":
    Standard:
      name: "ROW00016:"
      coefficients:
        - var_name: COL00006
          coefficient: -1
        - var_name: COL00012
          coefficient: -1
        - var_name: COL00017
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00017:":
    Standard:
      name: "ROW00017:"
      coefficients:
        - var_name: COL00007
          coefficient: -1
        - var_name: COL00013
          coefficient: -1
        - var_name: COL00018
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00018:":
    Standard:
      name: "ROW00018:"
      coefficients:
        - var_name: COL00008
          coefficient: -1
        - var_name: COL00014
          coefficient: -1
        - var_name: COL00019
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00019:":
    Standard:
      name: "ROW00019:"
      coefficients:
        - var_name: COL00017
          coefficient: -1
        - var_name: COL00020
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00020:":
    Standard:
      name: "ROW00020:"
      coefficients:
        - var_name: COL00018
          coefficient: -1
        - var_name: COL00021
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00021:":
    Standard:
      name: "ROW00021:"
      coefficients:
        - var_name: COL00019
          coefficient: -1
        - var_name: COL00022
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00022:":
    Standard:
      name: "ROW00022:"
      coefficients:
        - var_name: COL00020
          coefficient: -1
        - var_name: COL00023
          coefficient: 2
        - var_name: COL00024
          coefficient: 1
        - var_name: COL00025
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 170
  "ROW00023:":
    Standard:
      name: "ROW00023:"
      coefficients:
        - var_name: COL00021
          coefficient: -1
        - var_name: COL00023
          coefficient: 1
        - var_name: COL00024
          coefficient: 2
        - var_name: COL00025
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 130
  "ROW00024:":
    Standard:
      name: "ROW00024:"
      coefficients:
        - var_name: COL00016
          coefficient: 1
        - var_name: COL00022
          coefficient: -1
      sense: LessOrEqual
      rhs: 0
  "ROW00025:":
    Standard:
      name: "ROW00025:"
      coefficients:
        - var_name: COL00015
          coefficient: 1.1
        - var_name: COL00026
          coefficient: -1
      sense: Equal
      rhs: 0
  "ROW00026:":
    Standard:
      name: "ROW00026:"
      coefficients:
        - var_name: COL00016
          coefficient: 1
        - var_name: COL00026
          coefficient: 1
        - var_name: COL00027
          coefficient: -1
      sense: Equal
      rhs: 0
  "ROW00027:":
    Standard:
      name: "ROW00027:"
      coefficients:
        - var_name: COL00017
          coefficient: -1
        - var_name: COL00023
          coefficient: -1
        - var_name: COL00028
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00028:":
    Standard:
      name: "ROW00028:"
      coefficients:
        - var_name: COL00018
          coefficient: -1
        - var_name: COL00024
          coefficient: -1
        - var_name: COL00029
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00029:":
    Standard:
      name: "ROW00029:"
      coefficients:
        - var_name: COL00019
          coefficient: -1
        - var_name: COL00025
          coefficient: -1
        - var_name: COL00030
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00030:":
    Standard:
      name: "ROW00030:"
      coefficients:
        - var_name: COL00028
          coefficient: -1
        - var_name: COL00031
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00031:":
    Standard:
      name: "ROW00031:"
      coefficients:
        - var_name: COL00029
          coefficient: -1
        - var_name: COL00032
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00032:":
    Standard:
      name: "ROW00032:"
      coefficients:
        - var_name: COL00030
          coefficient: -1
        - var_name: COL00033
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00033:":
    Standard:
      name: "ROW00033:"
      coefficients:
        - var_name: COL00031
          coefficient: -1
        - var_name: COL00034
          coefficient: 2
        - var_name: COL00035
          coefficient: 1
        - var_name: COL00036
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 170
  "ROW00034:":
    Standard:
      name: "ROW00034:"
      coefficients:
        - var_name: COL00032
          coefficient: -1
        - var_name: COL00034
          coefficient: 1
        - var_name: COL00035
          coefficient: 2
        - var_name: COL00036
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 130
  "ROW00035:":
    Standard:
      name: "ROW00035:"
      coefficients:
        - var_name: COL00027
          coefficient: 1
        - var_name: COL00033
          coefficient: -1
      sense: LessOrEqual
      rhs: 0
  "ROW00036:":
    Standard:
      name: "ROW00036:"
      coefficients:
        - var_name: COL00026
          coefficient: 1.1
        - var_name: COL00037
          coefficient: -1
      sense: Equal
      rhs: 0
  "ROW00037:":
    Standard:
      name: "ROW00037:"
      coefficients:
        - var_name: COL00027
          coefficient: 1
        - var_name: COL00037
          coefficient: 1
        - var_name: COL00038
          coefficient: -1
      sense: Equal
      rhs: 0
  "ROW00038:":
    Standard:
      name: "ROW00038:"
      coefficients:
        - var_name: COL00028
          coefficient: -1
        - var_name: COL00034
          coefficient: -1
        - var_name: COL00039
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00039:":
    Standard:
      name: "ROW00039:"
      coefficients:
        - var_name: COL00029
          coefficient: -1
        - var_name: COL00035
          coefficient: -1
        - var_name: COL00040
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00040:":
    Standard:
      name: "ROW00040:"
      coefficients:
        - var_name: COL00030
          coefficient: -1
        - var_name: COL00036
          coefficient: -1
        - var_name: COL00041
          coefficient: 1
      sense: Equal
      rhs: 0
  "ROW00041:":
    Standard:
      name: "ROW00041:"
      coefficients:
        - var_name: COL00039
          coefficient: -1
        - var_name: COL00042
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00042:":
    Standard:
      name: "ROW00042:"
      coefficients:
        - var_name: COL00040
          coefficient: -1
        - var_name: COL00043
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00043:":
    Standard:
      name: "ROW00043:"
      coefficients:
        - var_name: COL00041
          coefficient: -1
        - var_name: COL00044
          coefficient: 1
      sense: LessOrEqual
      rhs: 0
  "ROW00044:":
    Standard:
      name: "ROW00044:"
      coefficients:
        - var_name: COL00042
          coefficient: -1
        - var_name: COL00045
          coefficient: 2
        - var_name: COL00046
          coefficient: 1
        - var_name: COL00047
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 170
  "ROW00045:":
    Standard:
      name: "ROW00045:"
      coefficients:
        - var_name: COL00043
          coefficient: -1
        - var_name: COL00045
          coefficient: 1
        - var_name: COL00046
          coefficient: 2
        - var_name: COL00047
          coefficient: 1.5
      sense: LessOrEqual
      rhs: 130
  "ROW00046:":
    Standard:
      name: "ROW00046:"
      coefficients:
        - var_name: COL00038
          coefficient: 1
        - var_name: COL00044
          coefficient: -1
      sense: LessOrEqual
      rhs: 0
  "ROW00047:":
    Standard:
      name: "ROW00047:"
      coefficients:
        - var_name: COL00037
          coefficient: 1.1
        - var_name: COL00048
          coefficient: -1
      sense: Equal
      rhs: 0
  "ROW00048:":
    Standard:
      name: "ROW00048:"
      coefficients:
        - var_name: COL00045
          coefficient: -0.8
        - var_name: COL00046
          coefficient: 0.1
        - var_name: COL00047
          coefficient: 0.15
      sense: LessOrEqual
      rhs: 0
  "ROW00049:":
    Standard:
      name: "ROW00049:"
      coefficients:
        - var_name: COL00045
          coefficient: 0.1
        - var_name: COL00046
          coefficient: -0.8
        - var_name: COL00047
          coefficient: 0.15
      sense: LessOrEqual
      rhs: 0
  "ROW00050:":
    Standard:
      name: "ROW00050:"
      coefficients:
        - var_name: COL00047
          coefficient: -1
        - var_name: COL00048
          coefficient: 1
      sense: LessOrEqual
      rhs: 0

